ccm-connector/src/main/java/com/sequenceiq/cloudbreak/ccmimpl/endpoint/DefaultServiceEndpointFinder.java

```diff
@@ -0,0 +1,119 @@
+package com.sequenceiq.cloudbreak.ccmimpl.endpoint;
+
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.sequenceiq.cloudbreak.ccm.endpoint.DirectServiceEndpointFinder;
+import com.sequenceiq.cloudbreak.ccm.endpoint.ServiceEndpoint;
+import com.sequenceiq.cloudbreak.ccm.endpoint.ServiceEndpointFinder;
+import com.sequenceiq.cloudbreak.ccm.endpoint.ServiceEndpointLookupException;
+import com.sequenceiq.cloudbreak.ccm.endpoint.ServiceEndpointRequest;
+
+/**
+ * Default service endpoint finder, that decides whether to return a direct service endpoint
+ * or try to look up a service endpoint via a gRPC call to minasshd. Uses a supplier to
+ * obtain the minasshd gRPC endpoint, to facilitate testing and to support dynamic creation
+ * of minasshd by the minasshd management service. Implements retrying, as well as caching
+ * the minasshd gRPC service endpoint finder to avoid unnecessary calls to the supplier.
+ */
+public class DefaultServiceEndpointFinder implements ServiceEndpointFinder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultServiceEndpointFinder.class);
+
+    /**
+     * A cache key for the minasshd endpoint finder.
+     */
+    private static final String MINASSHD_ENDPOINT_FINDER_CACHE_KEY = "minasshd.endpoint.finder.cache.key";
+
+    /**
+     * The default GRPC port.
+     */
+    private static final int DEFAULT_GRPC_PORT = 8984;
+
+    /**
+     * The cache expiration in seconds.
+     */
+    // JSA TODO make this configurable
+    private final int cacheExpirationInSeconds = 600;
```

Comment:
Should be static, and therefore CACHE_EXPIRATION_IN_SECONDS

Reply:
Done.